### ISSUE TITLE
WFA example error inside rules processing logic

### ISSUE THREAD

**Issue by prichards78** [braverock/quantstrat#62]
Hi, I'm trying to test the luxor walk forward example (`demo/luxor.8.walk.forward.R`) but the script fails with an error related to the rules processing logic, in the function `nextIndex` (defined inside `applyRules`):

```
[1] "=== testing param.combo 5 on 2002-10-27 16:30:00/2002-10-27 23:30:00"
  nFAST nSLOW
5     9    42
[1] "included indices:"
numeric(0)
Error in if (dindexNext < dindex[length(dindex)]) { : 
  argument is of length zero
```

I understand that this code is part of the *Dimension Reduction for Performance*, but as I'm new to quantstrat I couldn't make much sense of it.

I'm attaching the full log (generated with the command shown below) which also includes the session information (quantstrat 0.10.2, R 3.4.0 and Ubuntu 16.04).

```bash
Rscript -e "try(source('./demo/luxor.8.walk.forward.R')); cat('\n\n'); sessionInfo();" > walk_forward_error.txt 2>&1
```
[walk_forward_error.txt](https://github.com/braverock/quantstrat/files/1058601/walk_forward_error.txt)

**Comment by joshuaulrich** [comment]
Thanks for the report.  I can reproduce, and it looks like this was introduced in 293678f91c2f0882ec24d8d5a91d558203a7ee75.

**Comment by prichards78** [comment]
Hi Joshua, thanks for the detailed analysis. Following up on that, as you said, the bug seems to be introduced with the new [rule subsetting feature](https://github.com/braverock/quantstrat/commit/293678f91c2f0882ec24d8d5a91d558203a7ee75#diff-83dcc9f5b43a526c5d7e25badbb43930R352) of that commit. It's there where `dindex` is set to an empty vector (as reported by the print `numeric(0)`).

The line below that code section is setting [`dindex=1`](https://github.com/braverock/quantstrat/blob/293678f91c2f0882ec24d8d5a91d558203a7ee75/R/rules.R#L360) precisely for the case where `dindex` is empty, but it is not using the [`assign.dindex`](https://github.com/braverock/quantstrat/blob/293678f91c2f0882ec24d8d5a91d558203a7ee75/R/rules.R#L297) function (as the rest of the code does) and therefore the change is not being propagated to the code that actually causes the error. I'm speculating a bit here, because the code is quite hard to follow, there are a couple of big functions defined inside the `applyRules` function and `dindex` is handled in a different environment (` .Data`), I'm guessing to communicate it through these functions.

But even with that fix (adding `assign.dindex` after setting `dindex` to 1) I'm getting the same error you reported. Even by turning off the entire rule subsetting feature (by forcing `rule.subset` always to `NULL` and spiking the section that empties `dindex`) the result is the same. I tried to run this same example (`demo/luxor.8.walk.forward.R`) in versions previous to the problematic commit `293678f`, to have a reference for comparison, but I had other (apparently unrelated) errors that I couldn't clearly identify (`Error in runSum(x, n) : Invalid 'n'` inside `Calls: source ... applyIndicators -> do.call -> <Anonymous> -> runMean -> runSum`).

Is there a reference commit (something analogous to a release version) to run the example successfully and use it to better investigate the new error you reported?
